Add advanced network switch

RecursivePineapple opened the pull request:
This adds a new variant of the network switch. It's meant to be used in scenarios where you need to split data into many smaller streams (space elevator mainly). As such, the recipe will be balanced accordingly.

The advanced switch has a variable length (1 to 16 middle slices), and can accept or distribute computation to any optical hatch in its structure. Wallsharing is prevented by the structure to prevent computation duplication (only faces can have optical hatches).

I disabled maintenance on it so that it works more like an active transformer than a 'proper' multi.

<details><summary>Screenshots</summary>
<p>

![image](https://github.com/user-attachments/assets/4ac871d4-9ef3-451a-92a8-4e908d751f4d)
![image](https://github.com/user-attachments/assets/6756fcde-efa7-4039-b65c-00fe1c7727bf)

![2025-03-14_14 25 44](https://github.com/user-attachments/assets/ef593a3a-e8c8-4d91-800a-e9b3119c777c)
![2025-03-14_14 25 47](https://github.com/user-attachments/assets/ed3f2b75-1aeb-49ee-9af5-544917e6753c)
![2025-03-14_14 25 53](https://github.com/user-attachments/assets/e8cebef1-1f6f-4b93-a10c-a83e10e6061a)
![2025-03-14_14 25 55](https://github.com/user-attachments/assets/5affd3e7-1de2-4b75-8df5-b02db0ebf08e)
![2025-03-14_14 25 58](https://github.com/user-attachments/assets/1082b59b-947e-47e3-bc78-36d64dd51d2d)

</p>
</details> 

I'm not sure why the preview acts up when the length gets too long, survival construct doesn't place the hatches at all. There are also hatches everywhere in the preview due to an autoplace bug I can't fix easily. I looked at the AAL and I can't see what it does differently from what I do, it uses the same structurelib methods that I use here (`allowOnly` didn't work). These are fairly minor problems so I won't look into them unless they're problems with the structure wrapper.

<details><summary>Blockrenderer preview</summary>
<p>

![image](https://github.com/user-attachments/assets/07707e1b-ceae-4eb8-96e2-bfb4cf483d86)
![image](https://github.com/user-attachments/assets/5b881271-99e2-42b7-aa4a-3c85afaf2fc0)

</p>
</details> 

depends on: https://github.com/GTNewHorizons/GT5-Unofficial/pull/4067
depends on: https://github.com/GTNewHorizons/GT5-Unofficial/pull/4068

RecursivePineapple commented:
I removed the eu/t scaling by computation, and I removed the coolant mechanic. I also removed maintenance checks to bring it in line with active transformers. They're largely pointless by the time you need this since you'd have a huge stockpile of autotaping hatches anyways.

It seems like most people would rather deprecate the old switch and replace it with this one, so I changed it to fit the role better.

Glease commented:
at this point might as well just replace the old switch classes with new one. i see no reason to retain the old switch's code

RecursivePineapple commented:
That was my original plan, but I did a poll in #vote and a significant number of people like the old switch for some reason (around 20%). This PR is done, I just need to fix the recipes again and change the names again.

RecursivePineapple commented on src/main/java/tectech/thing/metaTileEntity/multi/MTENetworkSwitchAdv.java:
```diff
@@ -0,0 +1,392 @@
+            offset = offset.add(inc);
+        }
+
+        if (!structure.checkStructure(this, STRUCTURE_SHAPE_LAST, offset)) return false;
```
Unless I'm misunderstanding your message, that's a relatively simple problem to fix. If someone makes a structure whose middle and last slices are ambiguous (i.e. there is a block configuration that can match either), then the multi may not form. In something like the DT, you'd want to add a required air in the centre so that middle slices cannot match a final slice.

Now that I've thought about it, the network switch has a similar problem; if you replace all 4 edges on the last slice with output hatches, the network switch wouldn't have a way to determine whether the slice was a middle or last slice. The only solution to this would be to replace the internal 'bus' block (advanced computer casings currently) with another block, so that the last slice is no longer ambiguous. This would fix the problem because the last slice would require an advanced casing and not the new block, so an advanced casing surrounded by 4 hatches wouldn't match a middle slice.
I don't think this problem is big enough to require a new block, I highly doubt this would happen.

Something like the LSC is a good example, because the middle slices cannot contain glass in the centre 9 blocks. Only the last slice can contain glass. Since the last slice must only contain glass, the structure is well-defined and unambiguous.

Glease replied:
DT is not ambiguous. It's the way you are checking middle slice and depending on it to fail to start checking for final slice is problematic.

Let's say I have an HV energy hatch on the edge of last slice of a 3 block top DT. This would be the chain of event that happen under this current usage pattern

1. check 1st layer. all good
2. check 2nd layer using MIDDLE piece. all good
3. start checking 3rd layer (i.e. last layer) using MIDDLE piece.
4. It picks up the energy hatch on edge of 3rd layer. this is valid according to MIDDLE piece spec. This hatch got added to mEnergyHatches list.
5. It finds the center is not air but stainless casing, so it aborts WITHOUT resetting the hatch list.
6. Start checking 3rd layer (i.e. last layer) using LAST piece
7. It picks up the output hatch on edge of 3rd layer AGAIN. this is valid according to LAST piece spec. This hatch got added to mOuputHatches list AGAIN.
8. 3rd layer is found to be valid according to LAST piece. 
9. checkMachine() see it that everything is in place, the machine is now formed, **BUT with the last output hatch on last layer being duplicated in mEnergyHatches list**

Now because the mEnergyHatches now have two instances of the same actual energy hatch, which is a case that was never covered before, no deduplication would ever be done. The getMaxInputVoltage will now see two HV energy hatch in the list and start overclock it to EV, but you can never supply enough energy to this hatch, so the machine will powerfail regardless of what player do. The only option is to move energy hatch to another layer or behind the center block.

Glease replied:
In adv switch's case, you simply disallowed any hatch on the edge of last slice, so it cannot get into the aforementioned situation

Glease replied:
Actually, now that I see you do allow hatches on edges (just not data hatches), but adv switch does not derive its tier from energy hatch anyway, and tectech multiblocks has a different algorithm for draining energy from hatches, so it won't really cause any symptoms that player can see, but this is still a problem

e.g. this build leads to this mEnergyHatches
![图片](https://github.com/user-attachments/assets/dab2f28a-4fa6-4c37-a080-02ed07215d27)
![图片](https://github.com/user-attachments/assets/a7c81455-c966-4ad2-b5ee-d3eefb366bd0)

RecursivePineapple replied:
Hmm, I see what you mean. When you check a piece, it adds hatches regardless of whether the piece is fully formed. I'm not sure there's a robust, generic solution for this that doesn't require a rewrite of the hatch mechanism. Hatch adders will add a hatch to its list regardless of whether the structure piece is correct, so there isn't a way to tell a hatch adder to validate a hatch without adding it.

The hacky solution is to figure out some rule specific to the structure so that you can pre-determine which piece you're checking, but that's not portable and you shouldn't have to do that.

The best solution would be to add a `boolean simulate` param to hatch adders and do a simulated hatch pass, then a real hatch pass once the structure piece is known to be correct. When the param is true, adders should only validate the hatch but shouldn't modify any state.

I think this fix belongs in another PR since it's fundamental to the structure system and not something I've introduced here.

Glease replied:
I'd like a comment here stating this issue, so people don't blindly copy over existing "working" code without knowing this quirk.

RecursivePineapple replied:
I've added a comment and added a hack to mitigate these problems for now. When I have time I'll make a proper generic solution.

RecursivePineapple commented:
Alright, I've undeprecated the old switch and renamed the two to match their new roles. The old one is now called the weighted switch, while the new one is called the static switch. I also improved the old switch's tooltip and cleaned up its code a bit.

Testing done:
- Build a switch with several output hatches and feed data into it
- Put 100% of the computation into one of the hatches
- Split the data 50/50 with three hatches and make sure it works
- Make sure input hatches can only replace advanced casings

I still need to update the coremod PR to add a new conversion recipe. I won't make it anything fancy, it'll just be a crafting recipe.

![2025-04-27_16 48 39](https://github.com/user-attachments/assets/5733ec5c-27eb-41c0-99f4-c13e5aa4a37e)
![2025-04-27_16 48 40](https://github.com/user-attachments/assets/acf401bf-cfe5-4d82-9f92-e8f78d2c8222)

Glease reviewed (commented):
The rest of the code and new changes looks fine. Since the issues with repeated slices does not really cause a problem in adv switch specifically, I'd give my approval here as soon as the requested comment is added
